code-collab-app/src/storage/peer_data.js

```diff
@@ -68,71 +110,172 @@ const peer_data = observable({
             });
         });
     },
 
-    upload_document(d) {
-        this.doc_data = d;
+    handle_data_from_peers(data, dataConnection){
+        // Received a get doc request
+        if(data.message_type && data.message_type === GET_DOC_REQ){
+            console.log('received a get doc data request from another peer, sending data...');
+            dataConnection.send({
+                message_type: GET_DOC_RES,
+                content: this.doc_data
+            });
+        }
+        // Received a get doc response
+        else if(data.message_type && data.message_type === GET_DOC_RES && data.content){
+            console.log('received doc data from another peer');
+            this.doc_data.push(...data.content);
+        }
+        // Received a cell update
+        else if(data.message_type && data.message_type === CELL_UPDATE && data.content){
+                if(data.content.index >= 0){
+                    this.doc_data[data.content.index] = data.content.value;
+                }
+        }
+        else{
+            // TODO -- do something more with this data
+            console.log(data);
+        }
+    },
+
+    handle_error_from_peers(error){
+        console.log(error);
+    },
+
+    handle_peer_disconnect(data_connection){
+        console.log('a peer has been disconnected');
+        // remove the disconnected peer
+        this.session_peers = this.session_peers.filter(peerId => peerId !== data_connection.peer);
+        this.session_peers_conn = this.session_peers_conn.filter(conn => conn.peer !== data_connection.peer);
+    },
+
+    upload_document(new_data) {
+        this.doc_data.push(...new_data);
+    },
+
+    add_new_cell(cell_contents){
+        this.doc_data.push(cell_contents);
+        this.cell_locked.push(false);
+        this.increment_cell_count();
+        this.send_cell_update(this.doc_data.length - 1);
+    },
+
+    update_existing_cell(index, value){
+        this.doc_data[index] = value;
+    },
+
+    set_current_cell(index){
+        this.current_cell = index;
+    },
+
+    increment_cell_count(){
+        this.cell_count += 1;
+    },
+
+    set_cell_locked(index, value){
+        this.cell_locked[index] = value;
+    },
+
+    send_cell_update(key){
+        // broadcast to cell
+        console.log('sending the cell update to all the peers');
+        if(this.session_peers_conn !== null){
+            this.session_peers_conn.forEach(peer_conn => {
+                peer_conn.send({
+                    message_type: CELL_UPDATE,
+                    content: {
+                        index: key,
+                        value: this.doc_data[key]
+                    }
+                });
+            });
+        }
     },
 
-    // Get document data from the tracker if no peer active for the document
-    get_document_data(session_id) {
-        this.tracker.emit('get_doc_data', session_id);
+    get_current_cell(){
+        return this.current_cell;
```

Comment:
Just FYI. All the variables in `peer_data` store are accessible in the react component once its injected, so we don't really need getter/setter functions unless its useful to hide a bunch of calls to other peers.

E.G. 
- To get current cell in editor page we could use: `this.props.peer_data.current_cell`
- To change current cell in editor page we could use: `this.props.peer_data.current_cell = 1`

Reply:
Yeah I might remove all of this later on. I kept it in because we were using the local states. But I think I will remove most of them with an exception of the few where a lot of stuff is changed, that I will leave in the peer_data store to keep the mess out of editor page.